inline project env files in worker

james-elicx opened the pull request:
**Changes**

- Extracts env vars from .env files.
- Chucks them into a .env.mjs that is dynamically imported in the worker.

vicb commented on packages/cloudflare/src/cli/build/utils/extract-project-env-vars.ts:
```diff
@@ -0,0 +1,38 @@
+ * the env files at the root of the monorepo.
+ */
+export function extractProjectEnvVars(options: BuildOptions) {
+  const wranglerDevVars = readEnvFiles([".dev.vars"], options);
```
What about doing this at runtime?

A wrangler "environment variable" could be defined `.dev.vars`, `wrangler.toml` and the dashboard.

We would have to embed all the `env.*` files in the code, I don't think this would be an issue?

james-elicx replied:
My initial thought was we could use the readConfig function that is on the wrangler backlog somewhere, but this makes sense.

vicb commented on packages/cloudflare/src/cli/templates/worker.ts:
```diff
@@ -21,9 +21,13 @@ const cloudflareContextALS = new AsyncLocalStorage<CloudflareContext>();
 export default {
   async fetch(request, env, ctx) {
-    return cloudflareContextALS.run({ env, ctx, cf: request.cf }, async () => {
+    const combinedEnv = { ...__OPENNEXT_BUILD_TIME_ENV, ...env };
```
I think Next env vars should be on `process.env` ?

vicb commented:
A few thoughts (I'll do a deeper review later today)

I find "build time" to be confusing. Looking at the [next docs](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables) there are 2 things: Non-`NEXT_PUBLIC_` and `NEXT_PUBLIC_` vars.

Does the current code support [referencing other vars](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#referencing-other-variables) - maybe add a test.

It looks like we should support `development`, `production` and `test`. It looks like get the env from `NODE_ENV` so maybe `applyBuildTimeEnv(env["NEXTJS_ENV"])` should actually be `applyBuildTimeEnv(env["NEXTJS_ENV"] ?? process.env.NODE_ENV)`

To be more consistent with how Open Next work, we might want to generate `.env.development.mjs`, ... in the build folder of open next and include the file from worker.ts (rather than ESBulding the file). That would be similar to [compiling the config](https://github.com/opennextjs/opennextjs-cloudflare/blob/132b75e4a8492ca4c3c74d3b3ec07f50862722f5/packages/cloudflare/src/cli/build/index.ts#L36)

james-elicx commented on packages/cloudflare/src/cli/templates/worker.ts:
```diff
@@ -34,6 +49,8 @@ export default {
         },
       });
 
+      await applyProjectEnvVars(env.NEXTJS_ENV ?? globalThis.process.env.NODE_ENV);
```
This was super weird. For some bizarre reason, NODE_ENV was requiring it be accessed through globalThis to obtain the value:

<img width="340" alt="image" src="https://github.com/user-attachments/assets/69f3ab3b-2ef3-47ac-bed8-1bc5361cbb4c" />
<img width="204" alt="image" src="https://github.com/user-attachments/assets/03ba1644-4c68-4bd8-a87d-e1c118b21527" />

vicb replied:
Weird!

What about we default to "production" if `env.NEXTJS_ENV === undefined`
As discussed, we should add a `.dev.vars` for new projects

james-elicx replied:
And ignore `process.env.NODE_ENV` altogether I assume then?

james-elicx replied:
I've created https://github.com/opennextjs/opennextjs-cloudflare/issues/189 for the `.dev.vars` file.

vicb replied:
> And ignore `process.env.NODE_ENV` altogether I assume then?

The thing is `process.env` should be empty on first request to a worker anyway

james-elicx commented:
> I find "build time" to be confusing. Looking at the [next docs](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables) there are 2 things: Non-`NEXT_PUBLIC_` and `NEXT_PUBLIC_` vars.

Removed the reference to 'build-time'.

> Does the current code support [referencing other vars](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#referencing-other-variables) - maybe add a test.

Yeah, that was actually why i used the newer dotenvx instead of dotenv :). Added a test.

> It looks like we should support `development`, `production` and `test`. It looks like get the env from `NODE_ENV` so maybe `applyBuildTimeEnv(env["NEXTJS_ENV"])` should actually be `applyBuildTimeEnv(env["NEXTJS_ENV"] ?? process.env.NODE_ENV)`
> 
> To be more consistent with how Open Next work, we might want to generate `.env.development.mjs`, ... in the build folder of open next and include the file from worker.ts (rather than ESBulding the file). That would be similar to [compiling the config](https://github.com/opennextjs/opennextjs-cloudflare/blob/132b75e4a8492ca4c3c74d3b3ec07f50862722f5/packages/cloudflare/src/cli/build/index.ts#L36)

Sorry I hadn't realised it was an intentional part of the design. I've undone that change. Instead, it's not putting all of them into an `.env.mjs` file, exporting each mode. I originally did `.env.${mode}.mjs`, but this was causing runtime issues with variable-based dynamic imports due to wrangler's bundle step, so i couldn't do that unfortunately.

james-elicx commented on packages/cloudflare/src/cli/build/patches/investigated/copy-package-cli-files.ts:
```diff
@@ -6,7 +6,7 @@ import type { BuildOptions } from "@opennextjs/aws/build/helper.js";
 
 /**
- * Copies the template files present in the cloudflare adapter package into the standalone node_modules folder
+ * Copies the template files present in the cloudflare adapter package into the standalone node_modules folder.
```
```suggestion
 * Copies the template files present in the cloudflare adapter package into the standalone node_modules folder
```

vicb commented on packages/cloudflare/src/cli/build/patches/investigated/copy-env-files.ts:
```diff
@@ -0,0 +1,18 @@
+ * Copies the values extracted from the project's env files to the output directory for use in the worker.
+ */
+export function copyEnvFiles(options: BuildOptions) {
+  ["production", "development", "test"].forEach((mode) =>
```
it looks like we need the `.local` version for `wrangler dev` too
ref https://nextjs.org/docs/app/building-your-application/configuring/environment-variables#environment-variable-load-order

james-elicx replied:
it's extracting the local ones in the `extractProjectEnvVars` function

vicb commented on packages/cloudflare/src/cli/build/utils/extract-project-env-vars.ts:
```diff
@@ -0,0 +1,34 @@
+import { parse } from "@dotenvx/dotenvx";
+import type { BuildOptions } from "@opennextjs/aws/build/helper.js";
+
+function readEnvFiles(fileNames: string[], { monorepoRoot, appPath }: BuildOptions) {
```
nit: maybe it would be clearer to separate reading the file and merging the values in 2 different functions?

Also a comment to indicate the merge order would we helpful

vicb replied:
Maybe this can be `readEnvFile` (single file) and you use `.map()` - up to you

vicb commented on packages/cloudflare/src/cli/build/utils/extract-project-env-vars.ts:
```diff
@@ -0,0 +1,34 @@
+ * In a monorepo, the env files in an app's directory will take precedence over
+ * the env files at the root of the monorepo.
+ */
+export function extractProjectEnvVars(mode: string, options: BuildOptions) {
```
Doesn't that mean that we will always use the `.local` files while we only want to use them for `wrangler dev` but not in prod?

james-elicx replied:
Yeah. How come you wouldn't want them in production? Next.js would use those files. I suppose you're talking about if someone builds it locally and then deploys, but in that case, you could make the argument that we shouldnt include any .env files at all because technically they're all local to the machine. I'm not quite sure what else you could do here - open to suggestions though.

vicb replied:
@james-elicx I did a quick test and you are right, `.local` files are used if present, regardless if it is a prod or dev build. I was wrong to think that they would only be used in dev.

vicb commented on packages/cloudflare/src/cli/templates/worker.ts:
```diff
@@ -21,6 +21,21 @@ const cloudflareContextALS = new AsyncLocalStorage<CloudflareContext>();
   }
 );
 
+async function applyProjectEnvVars(mode = "production") {
```
Would doing this once at top-level remove the need for `process.env.__OPENNEXT_PROCESSED_ENV` ?

nit: maybe rename secrets to nextEnvVars? (those are not all secrets, i.e. `NEXT_PUBLIC_`)

james-elicx replied:
We can't do this at the top-level because of the desire to have the `NEXTJS_ENV` .dev.vars variable. If that's not needed, then I can move it to the top-level

vicb commented on packages/cloudflare/src/cli/build/utils/extract-project-env-vars.ts:
```diff
@@ -0,0 +1,35 @@
+ *
+ * The `NEXTJS_ENV` environment variable in `.dev.vars` determines the mode.
+ *
+ * Merged variables respect the following priority order.
```
Maybe add a link to https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#environment-variable-load-order

james-elicx commented on packages/cloudflare/src/cli/build/utils/extract-project-env-vars.ts:
```diff
@@ -0,0 +1,35 @@
+ * 2. `.env.local`
+ * 3. `.env.{mode}`
+ * 4. `.env`
+ *
```
```suggestion
 *
 * https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#environment-variable-load-order
 *
```

vicb commented on packages/cloudflare/src/cli/build/patches/investigated/copy-env-files.ts:
```diff
@@ -0,0 +1,18 @@
+import { extractProjectEnvVars } from "../../utils";
+
+/**
+ * Copies the values extracted from the project's env files to the output directory for use in the worker.
```
Maybe:

```suggestion
 * Compiles the values extracted from the project's env files to the output directory for use in the worker.
```

And rename the function `compilesEnvFiles` - similar to `compileOpenNextConfig`, `cvompiloeCache`, ...

vicb commented:
Do you think you can add a simple test, maybe to the api example app?

vicb reviewed (approved):
Awesome PR, thanks!

I would move the code out of patch*, I think it should be part of the build.

Approving the PR, you can pick whatever your prefer. 

🙏
